### DISCUSSION TITLE
Input (type text) with centered text

### DISCUSSION THREAD

**Discussion by xDarkDelux** [mikke89/RmlUi#354] [Q&A]
Hello, I reviewed a lot of examples on the Internet for html / css how to make input with centered text, but I could not do it in rml. Tell me, how can I make an input so that the text is always oriented to the center and not to the left? I would not want to change the RmlCore, and making your own custom element is not profitable. Can this be done in rml/rsss?

**Comment by mikke89** [comment]
Hi. I don't think our text input element supports that now. I don't see any way other than modifying core or making a custom text input element. If you do end up making something, pull requests are always welcome ;)

**Comment by xDarkDelux** [comment]
It is interesting that, for example, when a button is rendered (<button> tag), its name is immediately rendered next, with the #text tag, and with the <button> parent, with a call to OnRender in ElementText, and the same happens with the text type input, but the button text reacts to text-align:center, but the input does not :)

**Reply by mikke89** [reply]
Yeah, this is because the button uses normal layouting rules, while the text element does its own layouting to deal with selection, cursor and so on. In the latter case `text-align: center` would have to be supported manually.

**Comment by xDarkDelux** [comment]
And yet I decided to take the easy way and make changes to the core. This gave a great result and now the input responds to text-align:center. I will be glad if you point out the mistakes.

`void WidgetTextInput::OnRender()
{
	Vector2f offset = Vector2f(0.f, 0.f);

	if (parent->GetComputedValues().text_align == Style::TextAlign::Center)
	{
		int text_width = GetFontEngineInterface()->GetStringWidth(parent->GetFontFaceHandle(), text_element->GetText());
		offset.x = (parent->GetBox().GetSize().x - text_width) * 0.5;

		Vector2f text_position = parent->GetBox().GetPosition(Box::CONTENT);

		text_element->SetOffset(text_position + offset, parent);
		selected_text_element->SetOffset(text_position + offset, parent);
	}

	ElementUtilities::SetClippingRegion(text_element);

	Vector2f text_translation = parent->GetAbsoluteOffset() - Vector2f(parent->GetScrollLeft(), parent->GetScrollTop());
	selection_geometry.Render(text_translation + offset);

	if (cursor_visible &&
		!parent->IsDisabled())
	{
		cursor_geometry.Render(text_translation + cursor_position + offset);
	}
}`

**Reply by mikke89** [reply]
Good work!

I don't think it will work for multiline text-boxes, and I wonder if text selection/mouse interaction will not work correctly? In any case, if it works for your needs then all is good.

**Comment by xDarkDelux** [comment]
I don't know, haven't tested multiline text-boxes, but for single text-boxes everything works perfectly.
![Screenshot_1](https://user-images.githubusercontent.com/30740744/187033260-adc363b5-1419-49b2-a83f-7affc6d80570.png)
True, there is still a "problem" with the text box for the password, by default it draws stars too high, ideally they would be drawn in the center
